File: packages/app-floorplan/src/set-measurement-handler.ts
import { LengthMm } from "@toporoom/domain-floorplan";
import type {
  MeasurementKind,
  MeasurementSource,
  MeasurementTarget,
} from "@toporoom/domain-floorplan";
import type { DocumentStorePort } from "@toporoom/ports";
import { commit, loadDocument, type CommandResult } from "./document-io.js";
import type { GeometryRebuildPolicy } from "./rebuild-policy.js";

export interface SetMeasurementCommand {
  documentId: string;
  measurementId: string;
  kind: MeasurementKind;
  valueMm: number;
  source: MeasurementSource;
  instrumentId?: string;
  between?: readonly string[];
  target?: MeasurementTarget;
}

export class SetMeasurementHandler {
  constructor(
    private readonly store: DocumentStorePort,
    private readonly rebuild: GeometryRebuildPolicy,
  ) {}

  async execute(command: SetMeasurementCommand): Promise<CommandResult> {
    const document = await loadDocument(this.store, command.documentId);
    document.setMeasurement({
      id: command.measurementId,
      kind: command.kind,
      value: LengthMm.of(command.valueMm),
      source: command.source,
      instrumentId: command.instrumentId,
      between: command.between,
      target: command.target,
    });
    return commit(this.store, this.rebuild, document);
  }
}

File: core/src/app/set_measurement_handler.cpp
#include "toporoom/app/set_measurement_handler.hpp"

#include "toporoom/domain/floor_plan_document.hpp"
#include "toporoom/domain/length_mm.hpp"

namespace toporoom::app {

CommandResult SetMeasurementHandler::execute(const SetMeasurementCommand& command) {
  auto document = load_document(store_, command.document_id);
  domain::SetMeasurementProps props;
  props.id = command.measurement_id;
  props.kind = command.kind;
  props.value = domain::LengthMm::of(command.value_mm);
  props.source = command.source;
  props.instrument_id = command.instrument_id;
  props.between = command.between;
  props.target = command.target;
  document.set_measurement(std::move(props));
  return commit(store_, rebuild_, document);
}

}  // namespace toporoom::app
